File: ros_ws/src/helpers/utcoupe_metrics/src_py/utcoupe_metrics/metric_types/gauge.py
#!/usr/bin/env python

__author__ = "Gaëtan Blond"
__date__ = "20/03/20"

import rospy
from std_msgs.msg import Float64

class Gauge(object):
    def __init__(self, topic_name, start_value = 0.0):
        self.m_publisher = rospy.Publisher(topic_name, Float64, queue_size=10)
        self.m_value = start_value

    def inc(self, val_inc = 1.0):
        self.m_value += val_inc
    
    def dec(self, val_dec = 1.0):
        self.inc(-val_dec)

    def publish(self):
        self.m_publisher.publish(self.m_value)
    
    def set(self, val):
        self.m_value = val
    
    def set_to_current_time(self):
        self.m_value = rospy.get_rostime().to_nsec()

File: ros_ws/src/helpers/utcoupe_metrics/src/utcoupe_metrics/metric_types/gauge.cpp
#include "utcoupe_metrics/metric_types/gauge.h"

#include <std_msgs/Float64.h>

utcoupe_metrics::metric_types::Gauge::Gauge(ros::NodeHandle& nh, const std::string& topic_name, double start_value):
    m_value(start_value) {
    m_publisher = nh.advertise<std_msgs::Float64>(topic_name, 10);
}

void utcoupe_metrics::metric_types::Gauge::publish()
{
    std_msgs::Float64 msg;
    msg.data = m_value;
    m_publisher.publish(msg);
}

void utcoupe_metrics::metric_types::Gauge::setToCurrentTime()
{
    auto now = ros::Time::now();
    m_value = static_cast<double>(now.toNSec());
}

